Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.cpp

```diff
@@ -0,0 +1,501 @@
+                    auto& parentContainer = matchingRow->m_parentNode->m_childMatchState;
+                    parentContainer.erase(parentContainer.begin() + (patchPath.end() - 1)->GetIndex());
+
+                    // update former parent's match state, as its m_matchingDescendents may have changed
+                    UpdateMatchState(matchingRow->m_parentNode);
+                }
+                else
+                {
+                    // removed node wasn't a row, so we need to re-cache the owning row's cached info
+                    CacheDomInfoForNode(sourceContents[rowPath], matchingRow);
+                    UpdateMatchState(matchingRow);
+                }
+            }
+            else if (operationIterator->GetType() == AZ::Dom::PatchOperation::Type::Replace)
+            {
+                // replace operations can change child values from or into rows. Handle all cases
+                auto existingMatchInfo = GetMatchNodeAtPath(patchPath);
+                if (patchPath == rowPath)
+                {
+                    // replacement value is a row, populate it
+                    PopulateNodesAtPath(patchPath);
+                    if (!existingMatchInfo)
+                    {
+                        // this value wasn't a row before, so its parent needs to recache its info
+                        auto parentNode = GetMatchNodeAtPath(patchPath)->m_parentNode;
+                        CacheDomInfoForNode(sourceContents[patchPath].ArrayPopBack(), parentNode);
+                        UpdateMatchState(parentNode);
+                    }
+                }
+                else
+                {
+                    // not a row now. Check if it was before
+                    if (existingMatchInfo)
+                    {
+                        // node being replaced was a row. We need to remove it from our graph
+                        auto& parentContainer = existingMatchInfo->m_parentNode->m_childMatchState;
+                        parentContainer.erase(parentContainer.begin() + (patchPath.end() - 1)->GetIndex());
+
+                        // update former parent's match state, as its m_matchingDescendents may have changed
+                        UpdateMatchState(existingMatchInfo->m_parentNode);
+                    }
+                    else
+                    {
+                        // replaced node wasn't a row, so we need to re-cache the owning row's cached info
+                        auto matchingRow = GetMatchNodeAtPath(rowPath);
+                        CacheDomInfoForNode(sourceContents[rowPath], matchingRow);
+                        UpdateMatchState(matchingRow);
+                    }
+                }
+
+            }
+            else if (operationIterator->GetType() == AZ::Dom::PatchOperation::Type::Add)
+            {
+                // <apm> test this! Does an add imply shifting all subsequent nodes back by 1, or are we given those messages?
```

Comment:
Add should imply pushing all nodes back by 1, I believe - I believe but am not 100% sure we generate only append-to-end add patches right now for Node children which lets you get away with this.

Reply:
cool, fixed this by adding a replace parameter to PopulateNodesAtPath since the caller should always know if it's an add or replace.